eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,498 @@
+-- hierarchies and 4 `PrevGovActionIds`.
+data PForest (f :: Type -> Type) era = PForest
+  { pfPParamUpdate :: !(f (GovPurposeId 'PParamUpdatePurpose era))
+  , pfHardFork :: !(f (GovPurposeId 'HardForkPurpose era))
+  , pfCommittee :: !(f (GovPurposeId 'CommitteePurpose era))
+  , pfConstitution :: !(f (GovPurposeId 'ConstitutionPurpose era))
+  }
+
+deriving instance
+  ( Show (f (GovPurposeId 'PParamUpdatePurpose era))
+  , Show (f (GovPurposeId 'HardForkPurpose era))
+  , Show (f (GovPurposeId 'CommitteePurpose era))
+  , Show (f (GovPurposeId 'ConstitutionPurpose era))
+  ) =>
+  Show (PForest f era)
+
+pfPParamUpdateL :: Lens' (PForest f era) (f (GovPurposeId 'PParamUpdatePurpose era))
+pfPParamUpdateL = lens pfPParamUpdate $ \x y -> x {pfPParamUpdate = y}
+
+pfHardForkL :: Lens' (PForest f era) (f (GovPurposeId 'HardForkPurpose era))
+pfHardForkL = lens pfHardFork $ \x y -> x {pfHardFork = y}
+
+pfCommitteeL :: Lens' (PForest f era) (f (GovPurposeId 'CommitteePurpose era))
+pfCommitteeL = lens pfCommittee $ \x y -> x {pfCommittee = y}
+
+pfConstitutionL :: Lens' (PForest f era) (f (GovPurposeId 'ConstitutionPurpose era))
+pfConstitutionL = lens pfConstitution $ \x y -> x {pfConstitution = y}
+
+pfrPParamUpdateL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'PParamUpdatePurpose era))
+pfrPParamUpdateL = pfPParamUpdateL
+
+pfrHardForkL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'HardForkPurpose era))
+pfrHardForkL = pfHardForkL
+
+pfrCommitteeL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'CommitteePurpose era))
+pfrCommitteeL = pfCommitteeL
+
+pfrConstitutionL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'ConstitutionPurpose era))
+pfrConstitutionL = pfConstitutionL
+
+pfhPParamUpdateL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'PParamUpdatePurpose era))
+pfhPParamUpdateL = pfPParamUpdateL
+
+pfhHardForkL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'HardForkPurpose era))
+pfhHardForkL = pfHardForkL
+
+pfhCommitteeL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'CommitteePurpose era))
+pfhCommitteeL = pfCommitteeL
+
+pfhConstitutionL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'ConstitutionPurpose era))
+pfhConstitutionL = pfConstitutionL
+
+deriving instance Era era => Eq (PForest PRoot era)
+deriving instance Era era => Ord (PForest PRoot era)
+deriving instance Era era => Generic (PForest PRoot era)
+deriving instance Era era => NoThunks (PForest PRoot era)
+deriving instance Era era => NFData (PForest PRoot era)
+deriving instance Era era => Default (PForest PRoot era)
+
+deriving instance Era era => Eq (PForest PHierarchy era)
+deriving instance Era era => Ord (PForest PHierarchy era)
+deriving instance Era era => Generic (PForest PHierarchy era)
+deriving instance Era era => NoThunks (PForest PHierarchy era)
+deriving instance Era era => NFData (PForest PHierarchy era)
+deriving instance Era era => Default (PForest PHierarchy era)
```

Comment:
```suggestion
deriving instance Eq (f (GovPurposeId 'PParamUpdatePurpose era)) => Eq (PForest f era)
deriving instance Ord (f (GovPurposeId 'PParamUpdatePurpose era)) => Ord (PForest f era)
deriving instance Generic (f (GovPurposeId 'PParamUpdatePurpose era)) => Generic (PForest f era)
deriving instance NoThunks (f (GovPurposeId 'PParamUpdatePurpose era)) => NoThunks (PForest f era)
deriving instance NFData (f (GovPurposeId 'PParamUpdatePurpose era)) => NFData (PForest f era)
deriving instance Default (f (GovPurposeId 'PParamUpdatePurpose era)) => Default (PForest f era)
```

Reply:
I don't think this is gonna work, because `r` isn't present on the right hand side.
What you'll probably need is `QuantifiedConstraints`:
```haskell
deriving instance (forall r. Eq (f (GovPurposeId (r :: GovActionPurpose) era)) => 
  Eq (PForest f era)
```